operators/distiller-pipeline/distiller_pipeline/__init__.py

```diff
@@ -0,0 +1,981 @@
+        return df_receivers, receiver_totals
+
+    def get_aggregator_dataframe_and_totals(
+        self,
+    ) -> tuple[pd.DataFrame, dict[str, int]]:
+        """Prepares a DataFrame and totals for aggregator information."""
+        aggregator_data = []
+        aggregator_totals = {}
+        for agg_id in sorted(self.aggregator_map.keys()):
+            aggs_info = self.aggregator_map[agg_id]
+            aggregator_totals[agg_id] = aggs_info.n_messages
+            for module in sorted(aggs_info.aggregators.keys()):
+                threads_dict = aggs_info.aggregators[module]
+                for thread_id in sorted(threads_dict.keys()):
+                    info = threads_dict[thread_id]
+                    aggregator_data.append(
+                        {
+                            # "Agg ID": agg_id, # Optional
+                            "Module": info.module,
+                            # "Thread ID": thread_id, # Optional
+                            "Status": messaging_status_to_string(info.status),
+                            # "No. Messages": info.n_messages,
+                            "Scan No.": info.scan_number,
+                        },
+                    )
+        df_aggregators = pd.DataFrame(aggregator_data)
+        return df_aggregators, aggregator_totals
+
+    def get_node_group_dataframe_and_totals(
+        self,
+    ) -> tuple[pd.DataFrame, dict[str, int]]:
+        """Prepares a DataFrame and totals for node group information."""
+        node_group_data = []
+        node_totals = {}
+        for node_key in sorted(self.node_map.keys()):
+            node_info = self.node_map[node_key]
+            node_totals[node_info.node_id] = node_info.n_messages
+
+            aggregated_groups: dict[tuple[int, int], dict[str, Any]] = defaultdict(
+                lambda: {
+                    "n_messages": 0,
+                    "status": MessagingStatus.IDLE,
+                    "scan_num": float("inf"),
+                },
+            )
+            for mpi_rank in sorted(node_info.node_group_thread_info.keys()):
+                short_id_dict = node_info.node_group_thread_info[mpi_rank]
+                for short_id in sorted(short_id_dict.keys()):
+                    thread_id_dict = short_id_dict[short_id]
+                    group_key = (mpi_rank, short_id)
+                    for thread_id in sorted(thread_id_dict.keys()):
+                        ng_thread_info = thread_id_dict[thread_id]
+                        aggregated_groups[group_key]["n_messages"] += (
+                            ng_thread_info.n_messages
+                        )
+                        aggregated_groups[group_key]["status"] = ng_thread_info.status
+                        aggregated_groups[group_key]["scan_num"] = min(
+                            aggregated_groups[group_key]["scan_num"],
+                            ng_thread_info.scan_number,
+                        )
```

Comment:
These heavily nested for loops are pretty hard to follow, refactoring into methods might help.

Reply:
will fix along with other refactorings